imap-codec/src/codec/encode.rs

```diff
@@ -2130,4 +2133,26 @@ mod tests {
             assert_eq!(encoder.collect::<Vec<_>>(), actions);
         }
     }
+
+    #[cfg(feature = "serde")]
+    #[test]
+    fn test_serialize_fragment() {
+        let fragments = [
+            Fragment::Line {
+                data: b"A LOGIN alice {2}\r\n".to_vec(),
+            },
+            Fragment::Literal {
+                data: b"\xCA\xFE".to_vec(),
+                mode: LiteralMode::Sync,
+            },
+            Fragment::Line {
+                data: b"\r\n".to_vec(),
+            },
+        ];
+
+        assert_eq!(
+            serde_json::to_string(&fragments).unwrap(),
```

Comment:
Thought: AFAIR, I included `serde_json` just to "show-off" some of `imap-codec`s capabilities to some Thunderbird people. But, I played with the thought to remove the `serde_json` dep in our project to spare us some compile time. Have you seen it used somewhere lese besides the `serde_json` example? Can we rewrite the test to not use `serde_json` (so that we don't start to rely on it too much?)

Reply:
Can we use [`serde-test`](https://crates.io/crates/serde_test)?

Reply:
Argh, let's not overthink it. I'm fine with keeping this in for now :-)

Reply:
As always, let's create a ticket and hope that a future AI will solve it.

Reply:
Replacing the dev-dep of `serde-json` with `serde-test` (or any other crate) sounds a bit unnecessary. (I have to check out `serde-test`, though.)
However, if there is a way to describe the tests without requiring an additional dependency, I would definitely go for it.

Reply:
`serde_test` has less complexity (e.g. less dependencies). IMO that's important for dev, too.

Reply:
I agree. Still, I feel that it's sometimes more important to quickly merge PRs than to get everything right. Open PRs need rebasing, CI, etc. They may block developers because a local branch already depends on the changes. Opening tickets for later™ is okay :-)

Haha, was just about to file a ticket but saw that Henning already did :-) #549

Reply:
> `serde_test` has less complexity (e.g. less dependencies). IMO that's important for dev, too.

Maybe, I'm using it not correctly yet, but currently, the reduced readibility in the tests bothers me more:

```rust
#[test]
fn test_serialize_fragment() {
let fragments = [
    Fragment::Line {
        data: b"A LOGIN alice {2}\r\n".to_vec(),
    },
    Fragment::Literal {
        data: b"\xCA\xFE".to_vec(),
        mode: LiteralMode::Sync,
    },
    Fragment::Line {
        data: b"\r\n".to_vec(),
    },
];

// using serde_json
assert_eq!(
    serde_json::to_string(&fragments).unwrap(),
    r#"[{"Line":{"data":[65,32,76,79,71,73,78,32,97,108,105,99,101,32,123,50,125,13,10]}},{"Literal":{"data":[202,254],"mode":"Sync"}},{"Line":{"data":[13,10]}}]"#
);

// using serde_test
assert_ser_tokens(
    &fragments,
    &[
        Token::Tuple { len: 3 },
        Token::StructVariant {
            name: "Fragment",
            variant: "Line",
            len: 1,
        },
        Token::Str("data"),
        Token::Seq { len: Some(19) },
        Token::U8(65),
        Token::U8(32),
        Token::U8(76),
        Token::U8(79),
        Token::U8(71),
        Token::U8(73),
        Token::U8(78),
        Token::U8(32),
        Token::U8(97),
        Token::U8(108),
        Token::U8(105),
        Token::U8(99),
        Token::U8(101),
        Token::U8(32),
        Token::U8(123),
        Token::U8(50),
        Token::U8(125),
        Token::U8(13),
        Token::U8(10),
        Token::SeqEnd,
        Token::StructVariantEnd,
        Token::StructVariant {
            name: "Fragment",
            variant: "Literal",
            len: 2,
        },
        Token::Str("data"),
        Token::Seq { len: Some(2) },
        Token::U8(202),
        Token::U8(254),
        Token::SeqEnd,
        Token::Str("mode"),
        Token::UnitVariant {
            name: "LiteralMode",
            variant: "Sync",
        },
        Token::StructVariantEnd,
        Token::StructVariant {
            name: "Fragment",
            variant: "Line",
            len: 1,
        },
        Token::Str("data"),
        Token::Seq { len: Some(2) },
        Token::U8(13),
        Token::U8(10),
        Token::SeqEnd,
        Token::StructVariantEnd,
        Token::TupleEnd,
    ],
);
```

About "less dependencies":
```
> cargo tree -p imap-codec -F serde --no-dedupe
...
[dev-dependencies]
├── serde_json v1.0.120
│   ├── itoa v1.0.11
│   ├── ryu v1.0.18
│   └── serde v1.0.204
│       └── serde_derive v1.0.204 (proc-macro)
│           ├── proc-macro2 v1.0.86
│           │   └── unicode-ident v1.0.12
│           ├── quote v1.0.36
│           │   └── proc-macro2 v1.0.86
│           │       └── unicode-ident v1.0.12
│           └── syn v2.0.70
│               ├── proc-macro2 v1.0.86
│               │   └── unicode-ident v1.0.12
│               ├── quote v1.0.36
│               │   └── proc-macro2 v1.0.86
│               │       └── unicode-ident v1.0.12
│               └── unicode-ident v1.0.12
└── serde_test v1.0.176
    └── serde v1.0.204
        └── serde_derive v1.0.204 (proc-macro)
            ├── proc-macro2 v1.0.86
            │   └── unicode-ident v1.0.12
            ├── quote v1.0.36
            │   └── proc-macro2 v1.0.86
            │       └── unicode-ident v1.0.12
            └── syn v2.0.70
                ├── proc-macro2 v1.0.86
                │   └── unicode-ident v1.0.12
                ├── quote v1.0.36
                │   └── proc-macro2 v1.0.86
                │       └── unicode-ident v1.0.12
                └── unicode-ident v1.0.12
```
Yes, we could get rid of `itoa` and `ryu`.

I'm fine going both ways but would argue for using one way for all serde tests. (Meaning updating the tests in `imap-tests` in the near future when going with `serde_test`.)